Documents/AnalyticsToolDesignDocument.md

```diff
@@ -0,0 +1,48 @@
+# VFS For Git: Analytics Tooling Design Document
+
+This document details the design of and plan for the **VFS4G analytics tool**. The tool aims to help users identify reasons why their enlistment may be slowing down and the `gvfs` commands running slowly. By collecting and reporting statistics, the power to uncover and hopefully fix problems as they arrive moves to the user, empowering them to use the software to its fullest potential.
+
+## Problem
+
+When users unintentionally hydrate too many files within their enlistment, commands like checkout and commit increase their running time by several orders of magnitude. This causes unacceptable losses in performance and greatly hinders users as they go about their daily development cycle. The primary metric of the overhydration is the number of placeholder files which have been pulled from the remote into the local system. Slowdowns may also stem from having many modified files or directories which increases the amount of work that git does under the hood. Additionally, if a local enlistment has been in use for a long time, the number of loose object may eventually reach the order of several thousand in which case that also contributes to the loss in performance.

```

Comment:
> When users unintentionally hydrate too many files within their enlistment, commands like checkout and commit increase their running time by several orders of magnitude.

A small clarification on this point.  The performance issues that we've seen from users are usually caused by one (or both) of the following:

- Having too many placeholders on disk
- Having too many modified/full files (i.e. the files in "modified paths")

Hydrating files (i.e. populating empty placeholders with their contents) doesn't impact the steps VFS4G will need to take when processing a command like `git checkout`:

1. Check if the placeholder file that's on disk needs to be updated/deleted
2. If the placeholder needs to be updated, update it.  The helper function that VFS4G calls does this by:
    1. Deleting the old placeholder
    2. Writing a new placeholder
3. Store the updated information about the placeholder in the placeholder list

You might hear people referring to the problem of "over hydration" but from a technical standpoint it's not the hydrating specifically that's the problem.  It's having too many files/folders in the working directory.

Reply:
That makes sense. I rewrote for clarity of these points both for others but also to keep myself in track:

"Over the lifetime of an enlistment, sometimes a user's working directory fills up with too many entries. Generally this means having too many placeholders on their disk (have been read and pulled from remote) or too many files which are fully on disk (have been modified). This causes unacceptable losses in performance and greatly hinders users as they go about their daily development cycle. Additionally, if a local enlistment has been in use for a long time, the number of loose object may eventually reach the order of several thousand in which case that also contributes to the loss in performance."

Reply:
The re-write looks good.  I think you could simplify this:

> too many files which are fully on disk (have been modified)

To:
> too many modified files

Unfortunately some of the terminology used in VFS4G and ProjFS is not super straightforward:
- "Full files" -> Regular/plain NTFS files 
- "Hydrated placeholder" or "placeholder with contents" -> File that has all of its contents on disk, but still has ProjFS placeholder metadata (i.e. reparse data) attached to it.

And so both placeholders and modified files can be "fully on disk" as far as their contents are concerned.

Reply:
Got it! Good catch, I didn't know that but I'm glad that I do now. I'm pretty certain I made several more instances of that mistake so I'll go through and clarify there too